src/schematools/contrib/django/factories.py

```diff
@@ -321,6 +324,45 @@ def __call__(self, field: DatasetFieldSchema, dataset: DatasetSchema) -> TypeAnd
         return field_cls, args, kwargs
 
 
+def remove_dynamic_models():
+    """Erase model caches for dynamically generated models.
+    This completely removes the models from the Django app registry.
+
+    If your own code also holds references to models, these need to be removed separately.
+    The :func:`is_dangling_model` function allows to check whether a model originated
+    from a previous factory invocation.
+    """
+    virtual_apps = [
+        name
+        for name, config in apps.app_configs.items()
+        if isinstance(config, app_config.VirtualAppConfig)
+    ]
+    for app_label in virtual_apps:
+        del apps.all_models[app_label]
+        del apps.app_configs[app_label]
+
+    # See if there are dynamic models registered in other apps, erase them too.
+    for app_label, app_models in apps.all_models.items():
+        dynamic_models = [
+            name for name, model in app_models.items() if issubclass(model, DynamicModel)
+        ]
+        for model_name in dynamic_models:
+            del app_models[model_name]
+            del apps.app_configs[app_label].models[model_name]
+
+    # This also clears FK caches of other models that may have foreign keys:
+    apps.clear_cache()
+
+    # Allow code to detect whether a model is still alive despite having cleared all caches.
+    global MODEL_CREATION_COUNTER
+    MODEL_CREATION_COUNTER += 1
```

Comment:
Not sure how the dynamic models cache works (per process, per thread, something else), but this will almost certainly require thread/process synchronization (locking)

Reply:
I assumed this is mainly added for testing? So one process/thread? Or is runtime behaviour also depending on this code?

Reply:
good point. In the dso-api we have locking in place. I didn't dare to add locking at the low-level at this point but maybe we should?

Reply:
```py
class LockedCounter:
    """A thread-safe counter that starts at 1."""

    def __init__(self):
        self.__count = 0
        self.__lock = threading.Lock()

    @property
    def count(self):
        with self.__lock:
            return self.__count

    def next(self):
        with self.__lock:
            self.__count += 1
            return self.__count
```

Reply:
The problem is that the lock should cover a larger chunk of code: both the model creation and cache deletion. Thus, cache access should also be blocked, which a simple `threading.Lock` won't fix.